File: Python/easy/190. Reverse Bits.py
"""
Reverse bits of a given 32 bits unsigned integer.
Example 1:
Input: 00000010100101000001111010011100
Output: 00111001011110000010100101000000
Explanation: The input binary string 00000010100101000001111010011100 represents the unsigned integer 43261596,
so return 964176192 which its binary representation is 00111001011110000010100101000000.
Example 2:
Input: 11111111111111111111111111111101
Output: 10111111111111111111111111111111
Explanation: The input binary string 11111111111111111111111111111101 represents the unsigned integer 42949
"""
class Solution:
    # @param n, an integer
    # @return an integer
    def reverseBits(self, n):
        s = bin(n)
        s = s[2 : len(s)]
        while len(s) != 32:
            s = '0' + s
        s = s[::-1]
        return int(s, 2)

File: Java/easy/190. Reverse Bits.java
public class Solution {
    // you need treat n as an unsigned value
    final int precompute[] = {0, 8, 4, 12, 2, 10, 6, 14, 1, 9, 5, 13, 3, 11, 7, 15};
    
    public int reverseBits(int n) {
        int maskLength = 4;
        int mask = 0xF;
        int result = 0;
        for (int i = 0; i < 8; i++) {
            result |= (precompute[(n >> (i * maskLength)) & mask] << ((8 - i - 1) * maskLength)); 
        }
        return result;
    }
}